WIP: Add E2E tests

leonardola opened the pull request:
Can only be reviewed after #2238 gets merged

This is a Work in Progress PR so we can get feedback on the flows we are writing

## Changes proposed in this Pull Request:
This PR proposes to add E2E tests to the repository as there were none before. This way we can catch more errors before release and depend less on manual tests. Right now we have setup the environment and started writing some test flow.

Things to be discussed: 
- In WC Pay repo all the test flows are stored in a single file and inside merchant and shopper objects. That way the file becomes big really fast. So I decided to split the flows by area. Right now we have a `stripeUPESettingsUtils` object to store UPE related settings flows. Later we will have objects for non UPE settings, Other settings, Checkout, Pay Order page, My account, etc. This way we can separate flows into smaller files and objects.

- What flows should be added and what flows are the most important ones. Discussion at pdkmC9-7i-p2



## Testing instructions
No testing instructions for now. This is just a feedback request

a-danae committed:
Set up the base WC e2e boilerplate

At this point, the docker environment is the default one, without specific plugins or options set up. The tests don't run because puppeteer isn't set.

a-danae committed:
Install puppeteer and jest-environment-node

a-danae committed:
Add default.json file for the e2e tests

It's copied from WooCommerce, as a sample. We have to update it but it makes the tests run.

a-danae committed:
Activate the Stripe plugin in the e2e install

a-danae committed:
Create separate specs for merchant and shopper e2e tests flows

This will allow us to better organize the tests.

a-danae committed:
Use config.get() to make the test fail

Simply adding a line that will throw an error. For debugging.

a-danae committed:
Update the config package to 3.3.3

The previous version was somewhat old and caused unexpected errors when using the package in the e2e tests.

a-danae committed:
Add config file for e2e tests

This is the file that's actually retrieved when using the config package within the e2e tests.

a-danae committed:
Add a file for setting up the e2e tests flows

It's currently empty, to serve as a base.

leonardola committed:
added function to open the upe settings page

leonardola committed:
Merge remote-tracking branch 'origin/add/e2e-tests' into add/e2e-tests

leonardola committed:
added util to activate UPE and open UPE settings page

leonardola committed:
added test to validate it is possible to activate upe

leonardola committed:
fixed name of utils

leonardola committed:
fixed name of utils and added assertion to UPE flow

leonardola committed:
fixed name of utils and added assertion to UPE flow

leonardola committed:
Merge remote-tracking branch 'origin/add/e2e-tests-utils' into add/e2e-tests-utils

leonardola committed:
now the clickButtonWithText can click on buttons with a matching aria-label

leonardola committed:
added test to deactivate UPE settings

leonardola committed:
added util to disable Upe

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

a-danae committed:
Rename e2e-tests scripts to match the format from WCPay

The previous names were the default ones from the e2e boilerplate. Changing them to match the format from WCPay for consistency.

a-danae committed:
Update example test to get rid of the deprecation warning in the logs

This is an example tests that's going to be removed. Just for having a neat run.

a-danae committed:
Remove unused test file

a-danae committed:
Merge branch 'develop' into add/e2e-tests

leonardola committed:
added input name to make testing possible

leonardola committed:
added test to activate and deactivate upe method

leonardola committed:
renamed settings to UPESettings to only hold UPE tests. Added utils for activating and deactivating a upe payment method

leonardola committed:
Merge branch 'add/e2e-tests' into add/e2e-tests-utils

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
Merge branch 'develop' into e2e-tests-leo

a-danae committed:
Use a classic 'if' instead of optional chaining within useEffect's callback

The only reason behind this change is that the introduced @woocommerce/e2e-environment package locks Eslint to 6.7.2 which doesn't support optional chaining, making the linting fails in these cases. We could go back to optional chaining once the Eslint version is back to one higher than 7.5.0.

a-danae committed:
Make the mocked @wordpress/a11y module virtual

The previous mock stopped working after introducing @woocommerce/e2e-environment. Making it virtual allows mocking its 'speak' function, making the tests work as expected and as they worked before introducing the e2e-environment package.

lovo-h commented:
> In WC Pay repo all the test flows are stored in a single file... That way the file becomes big really fast. So I decided to split the flows by area.

I agree that splitting the test flows into different files makes sense, in order to create more maintainable test files. As for splitting the test flow by area, I am not entirely sure about this approach but the basis for the decision makes sense. 

Although, one thing I want to avoid is having too many files because then, instead of a really big file, we will end up with a directory with a lot of small files. Which may also be detrimental. Thoughts?

> What flows should be added and what flows are the most important ones. Discussion at pdkmC9-7i-p2

I went ahead and summarized what flows we should prioritize here: pdkmC9-7i-p2#comment-196.

chickenn00dle committed:
Merge branch 'develop' into add/e2e-tests

cesarcosta99 commented on tests/e2e/specs/merchant/UPESettings.test.js:
```diff
@@ -0,0 +1,18 @@
+import { merchant } from '@woocommerce/e2e-utils';
```
I'd suggest renaming the file and use kebab case (`upe-settings.test.js`), so we can keep consistency across the JS file names.

cesarcosta99 commented on tests/e2e/specs/merchant/example.test.js:
```diff
@@ -0,0 +1,22 @@
+/**
```
Let's not forget to remove this file 😄

cesarcosta99 commented on tests/e2e/utils/UPESettings.js:
```diff
@@ -0,0 +1,106 @@
+const baseUrl = config.get( 'url' );
+
+const UPE_SETTINGS_PAGE =
+	baseUrl + 'wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe';
```
Perhaps we can consider moving this into `tests/e2e/constants/screens.js` file or something like that, to gather all the screens we'll be navigating over the tests.

cesarcosta99 commented on tests/e2e/utils/UPESettings.js:
```diff
@@ -0,0 +1,106 @@
+import config from 'config';
```
Same about the kebab case file name.

cesarcosta99 commented on tests/e2e/utils/UPESettings.js:
```diff
@@ -0,0 +1,106 @@
+	/**
+	 * Opens Upe settings page
+	 */
+	openSettingsPage: async () => {
```
I'd recommend explicitly export the functions instead exporting an object with function properties. That way we can import only the functions we need and declare and use them more easily.

cesarcosta99 commented on tests/e2e/utils/buttons.js:
```diff
@@ -0,0 +1,23 @@
+export const buttonsUtils = {
+	clickButtonWithText: async ( text, parent = '' ) => {
```
Same about explicitly exporting the functions.

cesarcosta99 reviewed (commented):
This is looking nice and it's very promising, can't wait to have it fully integrated into our test suite 😃 Thanks @a-danae and @leonardola for kicking this off!

I've left a few suggestions in a project structure level, they're mostly about being consistent while adding new code.

a-danae committed:
Remove clearAllMocks after each test for testing

Giving it a quick test on whether this makes a difference for the tests that fail in GH but pass locally.

a-danae committed:
Revert "Remove clearAllMocks after each test for testing"

This reverts commit cee017d0ab76968d5d54a7d8f5d2559e1e4e5a53.

a-danae committed:
Clear individual mocks on afterEach

Another attempt to solve the tests passing locally but failing on GH.

a-danae committed:
Revert "Clear individual mocks on afterEach"

This reverts commit eb9f40c5a70853c28408828fd9c3dcf130be4ab3.

a-danae committed:
Use __esModule for mocking @wordpress/a11y

Is this the reason why the tests fail on GH? Let's see.

a-danae committed:
Revert "Use __esModule for mocking @wordpress/a11y"

This reverts commit 9d2aeb475b83189bb0b1da89b0ba74b601666023.

a-danae committed:
Remove unused mock of @wordpress/ally

The speak function doesn't seem to be affecting the tests. Also an attempt to figure out why the tests ran from the CI fail.

a-danae committed:
Revert "Remove unused mock of @wordpress/ally"

This reverts commit 1cff042f2ae403adedc6b932bb8c770fe58b6fad.

leonardola committed:
Merge branch 'develop' into add/e2e-tests

leonardola committed:
Merge branch 'add/e2e-tests' into add/e2e-tests-utils

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
renamed file

leonardola committed:
renamed file and fixed activa/deactivate upe

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
fixed name of parameter

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
fixed import

leonardola committed:
added test for basic card on classic checkout

leonardola committed:
added util for adding product to cart, filling checkout fields and filling upe card field

leonardola committed:
added card to e2e config file

leonardola committed:
removed example test file

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
added settings preparation before running tests and teardown

leonardola committed:
added basic auth plugin to e2e tests setup

leonardola committed:
added method to reset store/stripe settings before running tests

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
added htaccess rewrite to e2e test setup

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
added test to checkout with sca card and save card on account page

leonardola committed:
added SCA credit card

leonardola committed:
added functions to save credit card via my account

leonardola committed:
removed unecessary file

leonardola committed:
added comments and function to handle 3d card authentication

leonardola committed:
removed stripe keys from test files

leonardola committed:
added keys stripe keys to default.js and added a readme for e2e tests

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
now all methods are exported individually instead of being inside of an object

leonardola committed:
now all methods are exported individually instead of being inside of an object

leonardola committed:
removed unecessary file

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
fixed error that would happen after changing the location of the stripe account keys

leonardola committed:
started to create a test to remove saved cards

leonardola committed:
added some more details to readme

leonardola committed:
added functions for removing saved payment methods

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
added settings preparation before running tests and teardown

leonardola committed:
created function to remove all saved payment methods

leonardola committed:
fixed typo

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
separated tests for checkout and my account into multiple files

leonardola committed:
added test to checkout with saved card

leonardola committed:
fixed typo on function name. Now payment methods are removed before running checkout tests

leonardola committed:
fixed function name

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
separated invalid cards test from valid cards test

leonardola committed:
added new test cards

leonardola committed:
added tests for blocks checkout

leonardola committed:
moved some files to proper place

leonardola committed:
added WooCommerce checkout blocks plugin

leonardola committed:
moved stuff to proper place and added blocks checkout utils

leonardola committed:
added util to create a blocks checkout page

leonardola committed:
added email to customer billing address

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
fixed missing dependency for test

leonardola committed:
removed screenshot function

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

a-danae committed:
Rollback puppeteer version to ^2.0.0

This is the puppeteer version used by WCPay and previously required by @automattic/puppeteer-utils, which is the package we rely on for e2e tests. The version we were using before made the js unit tests fail due to a function from an external package not being mocked as expected.

leonardola committed:
removed unecessary screenshot.

leonardola committed:
added function to reset state before running tests on classic checkout with invalid cards

a-danae committed:
Lock puppeteer version at 2.1.1

leonardola committed:
added tests for invalid cards on blocks checkout

a-danae committed:
Remove @woocommerce/api package

For testing.

a-danae committed:
Revert "Remove @woocommerce/api package"

This reverts commit e2907cfcaf844f1f4c02213316569cd83f63f6b3.

a-danae committed:
Revert "Make the mocked @wordpress/a11y module virtual"

This reverts commit 0af0c20153ada330334e017fcb6116d921cb0e52.

a-danae committed:
Update package-lock.json

leonardola committed:
added invalid credit cards for e2e tests

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
moved the stripe test keys config to a separate file and created a template for it

leonardola committed:
added git ignore for e2e test keys file

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

a-danae committed:
Update the package-lock.json file with the one from develop

There's an issue with the file that's built from scratch. This happens on the develop branch as well. Let's build things on top of the develop file for things to work and check that issue out separately.

a-danae committed:
Uninstall config, puppeteer, and jest-environment-node packages

These are installed as dependencies of other packages.

a-danae committed:
Revert "Use a classic 'if' instead of optional chaining within useEffect's callback"

This reverts commit 595f0550572cd0681674245d820c851e6a7144d8.

leonardola committed:
fixed error that would not allow to save a credit card when postcode field is required

leonardola committed:
fixed simple product name

leonardola committed:
Merge branch 'add/e2e-tests-utils' into e2e-tests-leo

leonardola committed:
Merge branch 'add/e2e-tests' into add/e2e-tests-utils

leonardola committed:
added dotenv package

leonardola committed:
now the stripe keys are set inside a .env file instead of a json file. This makes it easier to set them on github actions

a-danae committed:
Comment out e2e path to a shopper directory that doesn't exist

This is added in a different branch.

a-danae committed:
Merge branch 'add/e2e-tests' into add/e2e-tests-utils

a-danae committed:
Revert "Comment out e2e path to a shopper directory that doesn't exist"

This reverts commit 83a5e88c80e14fe34f360e16903117784780db9e.

a-danae committed:
Only use local.env for local environment variables on e2e tests

This, just to stay consistent with how it's done on WCPay.

leonardola committed:
Merge branch 'develop' into add/e2e-tests-utils

leonardola committed:
updated package-lock

lovo-h commented:
Adding the blocked label back in since the tests in this PR need to be updated to reflect the changes that the repo underwent since these tests were first written. 

@leonardola, thanks for letting me know that you will try to update these tests, hopefully, during the next cycle cooldown.

diegocurbelo commented:
Closing this PR in favor of the recent Playwright E2E implementation.
